### PR TITLE
Utils financiers

### PR DESCRIPTION
**Pull request by Mussuranae**
Change cities of shooting locations

### CODE REVIEW

File: libs/movie/src/lib/movie/+state/movie.model.ts [hunk 1]
```diff
@@ -95,13 +95,13 @@ export function createMovie(params: Partial<Movie> = {}): Movie {
...
     audience: createAudienceGoals(params.audience),
-    estimatedBudget: createRange<number>(params.estimatedBudget),
+    estimatedBudget: createRange<number>(params.estimatedBudget ? params.estimatedBudget : {}),
     languages: createLanguageKey(params.languages ? params.languages : {}),
     poster: params.poster ?? '',
     promotional: createMoviePromotional(params.promotional),
-    release: createReleaseYear(params.release),
-    runningTime: createRunningTime(params.runningTime),
-    shooting: createShooting(params.shooting),
+    release: createReleaseYear(params.release ? params.release : {}),
```
Comment (GrandSchtroumpf) [inline libs/movie/src/lib/movie/+state/movie.model.ts:102]: Prefer to set a default {} value : 
```typescript
function createReleaseYear(release: MovieRelease = {}) { ... }
```
Same for below.

File: libs/movie/src/lib/movie/+state/movie.model.ts [hunk 2]
```diff
@@ -289,6 +289,9 @@ export function createShooting(params: Partial<MovieShooting> = {}): MovieShooti
 
 export function createShootingDate(params: Partial<MovieShootingDate> = {}): MovieShootingDate {
   return {
+    completed: null,
```
Comment (GrandSchtroumpf) [inline libs/movie/src/lib/movie/+state/movie.model.ts:292]: don't set default value to null if it's not a potential value.

File: libs/movie/src/lib/movie/form/shooting-information/shooting-information.component.html [hunk 3]
```diff
@@ -107,19 +107,29 @@ <h6 class="mat-caption">Shooting Dates</h6>
       <h6 class="mat-caption">Shooting Locations</h6>
       <bf-form-list [form]="form.shooting.get('locations')" buttonText="Add a Location">
         <ng-template itemRef let-location>
-          {{ location.city }} - {{ location.country | translateSlug: 'TERRITORIES' }}
+          {{ location.country | translateSlug: 'TERRITORIES' }} -
+          <ng-container *ngFor="let city of location.cities; let isLast=last">{{ city | titlecase }}{{isLast ? '' : ', '}}</ng-container>
```
Comment (GrandSchtroumpf) [inline libs/movie/src/lib/movie/form/shooting-information/shooting-information.component.html:111]: I thihnk this should be working
```html
{{ location.cities.joint(', ') | titlecase }}
```
Check if there is no "join" pipe already somewhere.

File: libs/movie/src/lib/movie/form/shooting-information/shooting-information.component.ts [hunk 4]
```diff
@@ -44,12 +43,14 @@ export class MovieFormShootingInformationComponent implements OnInit, OnDestroy
   }
 
   enableForm() {
-    this.sub = this.disabledForm.valueChanges.pipe(
-      distinctUntilChanged(),
-      filter(value => !!value),
-    ).subscribe(value => this.activate(value));
+    this.sub = this.disabledForm.valueChanges
```
Comment (GrandSchtroumpf) [inline libs/movie/src/lib/movie/form/shooting-information/shooting-information.component.ts:46]: I don't see the point of doing that. Go back to previous state.

File: libs/movie/src/lib/movie/form/shooting-information/shooting-information.component.ts [hunk 5]
```diff
@@ -63,4 +64,24 @@ export class MovieFormShootingInformationComponent implements OnInit, OnDestroy
       }
     }
   }
+
+  public add(event: any, control: FormControl): void {
```
Comment (GrandSchtroumpf) [inline libs/movie/src/lib/movie/form/shooting-information/shooting-information.component.ts:68]: Can't you type the event ? Like MouseEvent or something like that

File: libs/movie/src/lib/movie/form/shooting-information/shooting-information.component.ts [hunk 6]
```diff
@@ -63,4 +64,24 @@ export class MovieFormShootingInformationComponent implements OnInit, OnDestroy
...
+    // Add new value to the array
+    state.push(event.value.trim());
+    control.setValue(state);
+
+    // Reset the input
+    event.input.value = '';
+  }
+
+  public remove(i: number, control: FormControl): void {
+    console.log(control);
```
Comment (GrandSchtroumpf) [inline libs/movie/src/lib/movie/form/shooting-information/shooting-information.component.ts:80]: Remove console.log

File: libs/movie/src/lib/movie/form/shooting-information/shooting-information.component.ts [hunk 7]
```diff
@@ -63,4 +64,24 @@ export class MovieFormShootingInformationComponent implements OnInit, OnDestroy
...
+    const value = control.value;
```
Comment (GrandSchtroumpf) [inline libs/movie/src/lib/movie/form/shooting-information/shooting-information.component.ts:81]: you don't need to store the value in a variable

File: libs/movie/src/lib/movie/pipes/shooting.pipe.ts [hunk 8]
```diff
@@ -31,9 +31,10 @@ export class ShootingLocationsPipe implements PipeTransform {
     if (!shooting.locations.length) {
       return 'No location provided';
     } else {
-      return shooting.locations.map(({ city, country }) => {
+      return shooting.locations.map(({ cities, country }) => {
         const territory = formatSlug(country, 'TERRITORIES');
-        return city ? `${territory} (${city})` : territory;
+        const allCities = cities.join();
```
Comment (GrandSchtroumpf) [inline libs/movie/src/lib/movie/pipes/shooting.pipe.ts:36]: I think you want to `join(', ')`

File: libs/movie/src/lib/movie/pipes/shooting.pipe.ts [hunk 9]
```diff
@@ -31,9 +31,10 @@ export class ShootingLocationsPipe implements PipeTransform {
     if (!shooting.locations.length) {
       return 'No location provided';
     } else {
-      return shooting.locations.map(({ city, country }) => {
+      return shooting.locations.map(({ cities, country }) => {
```
Comment (GrandSchtroumpf) [inline libs/movie/src/lib/movie/pipes/shooting.pipe.ts:34]: What about having a vertical list of countries with a list of chips for the cities ?
Check with Stéphanie

**Review by GrandSchtroumpf — CHANGES_REQUESTED** [review]

File: libs/movie/src/lib/movie/+state/movie.model.ts (hunk 2 above)
Comment (phpgeekfr) [inline libs/movie/src/lib/movie/+state/movie.model.ts:292]: This is how we should handle number initialization now ?

File: libs/movie/src/lib/movie/+state/movie.model.ts [hunk 10]
```diff
@@ -95,13 +95,13 @@ export function createMovie(params: Partial<Movie> = {}): Movie {
     ...params,
     banner: params.banner ?? '',
     audience: createAudienceGoals(params.audience),
-    estimatedBudget: createRange<number>(params.estimatedBudget),
+    estimatedBudget: createRange<number>(params.estimatedBudget ? params.estimatedBudget : {}),
```
Comment (phpgeekfr) [inline libs/movie/src/lib/movie/+state/movie.model.ts:98]: Maybe you can do `createRange<number>(params.estimatedBudget || {}),`

Comment (phpgeekfr) [reply libs/movie/src/lib/movie/+state/movie.model.ts:98]: But for example `createRange<D>(params: Partial<RawRange<D>> = {})` already have {} as default, so is it necessary ?

File: libs/movie/src/lib/movie/form/shooting-information/shooting-information.component.ts [hunk 11]
```diff
@@ -44,12 +43,14 @@ export class MovieFormShootingInformationComponent implements OnInit, OnDestroy
...
-    ).subscribe(value => this.activate(value));
+    this.sub = this.disabledForm.valueChanges
+      .pipe(
+        distinctUntilChanged(),
+        filter((value) => !!value)
+      )
+      .subscribe((value) => this.activate(value));
 
-    const active = this.keys.find(key => hasValue(this.form.shooting.value.dates[key]));
+    const active = this.keys.find((key) => hasValue(this.form.shooting.value.dates[key]));
```
Comment (phpgeekfr) [inline libs/movie/src/lib/movie/form/shooting-information/shooting-information.component.ts:53]: extra parenthesis  for (key) and (value) . Did you updated your linter ?

File: libs/movie/src/lib/movie/form/shooting-information/shooting-information.component.html (hunk 3 above)
Comment (Mussuranae) [reply libs/movie/src/lib/movie/form/shooting-information/shooting-information.component.html:111]: it works like that without pipe :)

File: libs/movie/src/lib/movie/form/shooting-information/shooting-information.component.ts (hunk 11 above)
Comment (Mussuranae) [reply libs/movie/src/lib/movie/form/shooting-information/shooting-information.component.ts:53]: Nope Oo'
I changed it

File: libs/movie/src/lib/movie/pipes/shooting.pipe.ts (hunk 8 above)
Comment (Mussuranae) [reply libs/movie/src/lib/movie/pipes/shooting.pipe.ts:36]: It's the default behavior, I saw that I don't need to add `', '` to get it correctly :)

File: libs/movie/src/lib/movie/+state/movie.model.ts (hunk 2 above)
Comment (Mussuranae) [reply libs/movie/src/lib/movie/+state/movie.model.ts:292]: No, I deleted it, it's an optional data so I can remove the initialization

File: libs/movie/src/lib/movie/+state/movie.model.ts (hunk 10 above)
Comment (Mussuranae) [reply libs/movie/src/lib/movie/+state/movie.model.ts:98]: It's not, I removed it ^^

File: libs/movie/src/lib/movie/pipes/shooting.pipe.ts (hunk 8 above)
Comment (GrandSchtroumpf) [reply libs/movie/src/lib/movie/pipes/shooting.pipe.ts]: No the default value is "," not ", ". The space is missing

**Review by GrandSchtroumpf — APPROVED** [review]

**Pull request merged**